tests/unit/test_charm.py

```diff
@@ -0,0 +1,335 @@
+        "config.KafkaConfig.set_server_properties"
+    ), patch(
+        "charm.KafkaCharm.ready_to_start", new_callable=PropertyMock, return_value=False
+    ), patch(
+        "snap.KafkaSnap.start_snap_service"
+    ) as patched_service:
+        harness.charm.on.start.emit()
+
+        patched_service.assert_not_called()
+
+
+def test_start_does_not_start_if_not_same_tls_as_zk(harness):
+    peer_rel_id = harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "enabled",
+        },
+    )
+    harness.update_relation_data(peer_rel_id, CHARM_KEY, {"sync-password": "mellon"})
+
+    with patch("auth.KafkaAuth.add_user"), patch("config.KafkaConfig.set_jaas_config"), patch(
+        "config.KafkaConfig.set_server_properties"
+    ), patch("snap.KafkaSnap.start_snap_service") as patched_service:
+        harness.charm.on.start.emit()
+
+        patched_service.assert_not_called()
+        assert isinstance(harness.charm.unit.status, BlockedStatus)
+
+
+def test_start_does_not_start_if_leader_has_not_set_creds(harness):
+    peer_rel_id = harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "enabled",
+        },
+    )
+    harness.update_relation_data(peer_rel_id, CHARM_KEY, {"sync-password": "mellon"})
+
+    with patch("config.KafkaConfig.set_jaas_config"), patch(
+        "config.KafkaConfig.set_server_properties"
+    ), patch("snap.KafkaSnap.start_snap_service") as patched_service:
+        harness.charm.on.start.emit()
+
+        patched_service.assert_not_called()
+        assert isinstance(harness.charm.unit.status, BlockedStatus)
+
+
+def test_start_blocks_if_service_failed_silently(harness):
+    peer_rel_id = harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "disabled",
+        },
+    )
+    harness.update_relation_data(peer_rel_id, CHARM_KEY, {"sync-password": "mellon"})
```

Comment:
you seem to do this several times, making it into a function will make your tests more clear